compositor_render/src/transformations/builtin/params/tiled_layout.rs

```diff
@@ -73,69 +75,98 @@ impl TiledLayoutParams {
         inputs_count: u32,
         rows_cols: &RowsCols,
         tile_size: Resolution,
-        output_resolution: Resolution,
+        tailed_layout_spec: &TailedLayoutSpec,
     ) -> Vec<BoxLayout> {
         let mut layouts = Vec::with_capacity(inputs_count as usize);
 
-        let tiles_width = tile_size.width as f32 * rows_cols.cols as f32;
-        let x_padding = (output_resolution.width as f32 - tiles_width) / 2.0;
-
-        let tiles_height = tile_size.height as f32 * rows_cols.rows as f32;
-        let y_padding = (output_resolution.height as f32 - tiles_height) / 2.0;
+        // Because scaled tails with padding and margin don't have to cover whole output frame,
+        // additional padding is distributed is distributed accordingly to alignment
+        let additional_y_padding = tailed_layout_spec.resolution.height as u32
+            - (tile_size.height as u32 + 2 * tailed_layout_spec.padding) * rows_cols.rows
+            - (tailed_layout_spec.margin * (rows_cols.rows + 1));
+
+        let (additional_top_padding, justified_padding_y) =
+            match tailed_layout_spec.vertical_alignment {
+                VerticalAlign::Top => (0.0, 0.0),
+                VerticalAlign::Center => (additional_y_padding as f32 / 2.0, 0.0),
+                VerticalAlign::Bottom => (additional_y_padding as f32, 0.0),
+                VerticalAlign::Justified => {
+                    let space = additional_y_padding as f32 / (rows_cols.rows + 1) as f32;
+                    (0.0, space)
+                }
+            };
 
-        for row in 0..(rows_cols.rows - 1) {
-            for col in 0..rows_cols.cols {
-                let top_left_corner = (
-                    x_padding + col as f32 * tile_size.width as f32,
-                    y_padding + row as f32 * tile_size.height as f32,
-                );
+        let mut top = additional_top_padding
+            + justified_padding_y
+            + tailed_layout_spec.padding as f32
+            + tailed_layout_spec.margin as f32;
```

Comment:
This and `left` don't have to be `mut`, but I found modifying it more readable than calculating it in every iteration